File: chart.cc
#include <sstream>
#include <stdexcept>
#include "chart.hh"
#include "utils.hh"

namespace ndatk
{
  // Canonicalize: convert z000 to z; pass other sza unchanged
  int canonicalize(int sza)
  {
    if (sza % 1000 == 0)  
      return sza/1000;          // Z000 -> Z
    else 
      return sza;
  }

  using namespace std;

  // Construct Chart of the Nuclides from input stream
  NuclidesChart::NuclidesChart(istream &s)
  {
    string line;
    enum states {START, TABLE, CHART}; // Parser states
    states state = START;

    int sza;
    ElementData e;
    NuclideData n;

    // function getlogicalline: get line with continuation
    while (getline(s, line)) {
      trim(line);      // remove leading, trailing whitespace
      if (line.size() > 0 && line[0] != '#') {
        if (starts_with_nocase(line, "NAME:")) {
          getline(s, line);
          trim(line);
          name = line;
        } else if (starts_with_nocase(line, "NAME:")) {
          getline(s, line);
          trim(line);
          name = line;
        } else if (starts_with_nocase(line, "NAME:")) {
          getline(s, line);
          trim(line);
          name = line;
        } else if (starts_with_nocase(line, "PERIODIC_TABLE:")) {
          state = TABLE;
        } else if (starts_with_nocase(line, "CHART_OF_THE_NUCLIDES:")) {
          state = CHART;
        } else if (state == TABLE) {
          // function split: tokenize line into words
          istringstream iline(line);
          iline >> sza >> e.symbol >> e.at_wgt >> e.name;
          element.push_back(e);
        } else if (state == CHART) {
          // function split: tokenize line into words
          istringstream iline(line);
          iline >> sza >> n.awr >> n.abundance >> n.half_life;
          nuclide.insert(Nuclide_map::value_type(sza, n));
        }
      }
    }
  }

  // Return string value based on key
  string NuclidesChart::get_string_val(string_val key) const
  {
    switch (key) {
    case LIBRARY:
      return name;
      break;
    case DATE:
      return date;
      break;
    case INFO:
      return info;
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }

  // Return int value based on key
  int NuclidesChart::get_int_val(int_val key) const
  {
    switch (key) {
    case NUM_ELEMENTS:
      return element.size();
      break;
    case NUM_NUCLIDES:
      return nuclide.size();
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }
        
  // Return string value based on key and index
  string NuclidesChart::get_string_val_n(string_val_n key, int sza) const
  {
    int n = canonicalize(sza);             // Canonicalize SZA
    switch (key) {
    case SYMBOL:
      return element.at(n).symbol;
      break;
    case NAME:
      return element.at(n).name;
      break;
    default:
      throw out_of_range("Key not found!");
    }
  }

  // Return double value based on key and index
  double NuclidesChart::get_float_val_n(float_val_n key, int sza) const
  {
    int n = canonicalize(sza);
    if (n < element.size()) {  // Element data
      switch(key) {
      case AT_WGT:
        return element.at(n).at_wgt;
        break;
      case AWR:
        return element.at(n).at_wgt / element.at(0).at_wgt;
        break;
      default:
        throw out_of_range("Key not found!");
      }
    } else {
      NuclideData d = map_at(nuclide, n);
      switch (key) {            // Nuclide data
      case  AT_WGT:
        return d.awr * element.at(0).at_wgt;
        break;
      case AWR:
        return d.awr;
        break;
      case ABUNDANCE:
        return d.abundance;
        break;
      case HALF_LIFE:
        return d.half_life;
        break;
      default:
        throw out_of_range("Key not found!");
      }
    }
  }
}

File: chart.py
#! /usr/bin/env python
"""Chart of the Nuclides"""

__author__ = "Mark G. Gray"
__copyright__ = "Copyright 2013"
__credits__ = ["Mark G. Gray"]

__version__ = "1.0"
__maintainer__ = "Mark G. Gray"
__email__ = "gray@lanl.gov"
__status__ = "Development"

class NuclidesChart(object):
    """Chart of the Nuclides"""
    def __init__(self, f):
        """Parse f for peroidic table and chart of the nuclides"""
        self.elements = []
        self.nuclides = {}
        state = ""
        for line in f:
            s = line.strip()    # Trim whitespace
            if len(s) > 0 and not s.startswith('#'):
                if s.startswith("[PERIODIC TABLE]"):
                    state = "pt"
                elif s.startswith("[CHART OF THE NUCLIDES]"):
                    state = "cn"
                elif state == "pt":
                    sf = s.split()
                    self.elements.append((sf[1], float(sf[2]), sf[3]))
                elif state == "cn":
                    sf = s.split()
                    self.nuclides[int(sf[0])] = (float(sf[1]), float(sf[2]),
                                                 float(sf[3]))

    def get_string_val_n(self, key, n):
        """return string value by key and index"""
        if key == "SYMBOL":
            return self.elements[n][0]
        elif key == "NAME":
            return self.elements[n][2]
        else:
            raise KeyError, "%s invalid key!" % key

    def get_float_val_n(self, key, n):
        """return float value by key and index"""
        if key == "AT_WGT":
            if n < len(self.elements):
                return self.elements[n][1]
            else:
                return self.nuclides[n][0] * self.elements[0][1]
        elif key == "AWR":
            if n < len(self.elements):
                return self.elements[n][1] / self.elements[0][1]
            else:
                return self.nuclides[n][0]
        elif key == "ABUNDANCE":
            return self.nuclides[n][1]
        elif key == "HALF_LIFE":
            return self.nuclides[n][2]
        else:
            raise KeyError, "%s invalid key!" % key

        
if __name__ == "__main__":
    import sys
    a = NuclidesChart(open(sys.argv[1]))

    for i in range(10):
        print i, a.get_string_val_n("SYMBOL", i), a.get_float_val_n("AWR", i)

        
    print a.get_float_val_n("AT_WGT", 1001)
    print a.get_float_val_n("ABUNDANCE", 17035)
    print a.get_float_val_n("HALF_LIFE", 92235)
